cpp/src/arrow/csv/chunker.cc

```diff
@@ -63,12 +63,18 @@ class Lexer {
       case IN_FIELD:
         goto InField;
       case AT_ESCAPE:
+        // will never reach here if escaping = false
+        // just to hint the compiler to remove dead code
+        if (!escaping) return nullptr;
         goto AtEscape;
       case IN_QUOTED_FIELD:
+        if (!quoting) return nullptr;
         goto InQuotedField;
```

Comment:
Compiler is not able to figure out this code path is impossible when template augment `quoting` is `false`.
Adding this hint makes compiler eliminating unnecessary dead code.

Reply:
The way I validate above assumption is a bit rough.
I added below code snippet after [InQuotedField](https://github.com/apache/arrow/blob/de7cc1e4820e96531aae3b7f3fc9105f7b3a5f3a/cpp/src/arrow/csv/chunker.cc#L124)
```
InQuotedField:
  if (!quoting) {
    __asm volatile ("pause" ::: "memory");
    __asm volatile ("pause" ::: "memory");
    __asm volatile ("pause" ::: "memory");
  }
```
Logically, these three "pause" instruction should not be available.
Without this patch, I can find the three continuous pause in disassembled libarrow.so. With this patch, they are gone.